flink-ml-core/src/main/java/org/apache/flink/ml/common/datastream/DataStreamUtils.java

```diff
@@ -288,13 +291,20 @@ public static <IN, ACC, OUT> DataStream<OUT> aggregate(
     public static <T> DataStream<T> sample(DataStream<T> input, int numSamples, long randomSeed) {
         int inputParallelism = input.getParallelism();
 
-        return input.transform(
+        // In a worst-case scenario, the data partition with the greatest number of elements has
+        // `inputParallelism` additional elements compared to the one with the fewest elements even
+        // after `rebalance` performed. Therefore, additional elements are sampled from each
+        // partition in case some partitions has insufficient elements.
+        int firstRoundNumSamples =
```

Comment:
nit: The max difference of number of data points in each partition after calling `rebalance` is `inputParallism`. 
As a result, extra `inputParallelism` data points are sampled for each partition in the first round.